tsdb/shard.go

```diff
@@ -1786,154 +1830,245 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
+		Logger:        logger,
+		goodSize:      int64(0),
+	}
+	fs.changeMgr.Wg.Add(1)
+	go fs.changeMgr.SaveWriter()
 }
 
-func (w *MeasurementFieldSetWriter) Close() {
-	if w != nil {
-		close(w.writeRequests)
-		w.wg.Wait()
+func (fscm *measurementFieldSetChangeMgr) Close() {
+	if fscm != nil {
+		close(fscm.writeRequests)
+		fscm.Wg.Wait()
 	}
 }
 
-func (fs *MeasurementFieldSet) Save() error {
-	return fs.writer.RequestSave()
+func (fs *MeasurementFieldSet) Save(changes FieldChanges) error {
+	return fs.changeMgr.RequestSave(changes)
 }
 
-func (w *MeasurementFieldSetWriter) RequestSave() error {
+func (fscm *measurementFieldSetChangeMgr) RequestSave(changes FieldChanges) error {
 	done := make(chan error)
-	wr := writeRequest{done: done}
-	w.writeRequests <- wr
+	fscm.writeRequests <- writeRequest{errorReturn: done, changes: changes}
 	return <-done
 }
 
-func (fs *MeasurementFieldSet) saveWriter() {
-	defer fs.writer.wg.Done()
-	// Block until someone modifies the MeasurementFieldSet and
-	// it needs to be written to disk.
-	for req, ok := <-fs.writer.writeRequests; ok; req, ok = <-fs.writer.writeRequests {
-		fs.writeToFile(req)
+func (fscm *measurementFieldSetChangeMgr) SaveWriter() {
+	defer fscm.Wg.Done()
+	// Block until someone modifies the MeasurementFieldSet, and
+	// it needs to be written to disk. Exit when the channel is closed
+	for wr, ok := <-fscm.writeRequests; ok; wr, ok = <-fscm.writeRequests {
+		fscm.writeChangesToFile(wr)
 	}
 }
 
-// writeToFile: Write the new index to a temp file and rename when it's sync'd
-func (fs *MeasurementFieldSet) writeToFile(first writeRequest) {
-	var err error
-	// Put the errorChannel on which we blocked into a slice to allow more invocations
-	// to share the return code from the file write
-	errorChannels := []errorChannel{first.done}
-	defer func() {
-		for _, c := range errorChannels {
-			c <- err
-			close(c)
-		}
-	}()
-	// Do some blocking IO operations before marshalling the in-memory index
-	// to allow other changes to it to be queued up and be captured in one
-	// write operation, in case we are under heavy field creation load
+// WriteToFileNoLock: Write the new index to a temp file and rename when it's sync'd
+// This should only be called when writes are blocked.
+func (fs *MeasurementFieldSet) WriteToFileNoLock() error {
 	path := fs.path + ".tmp"
 
 	// Open the temp file
 	fd, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_EXCL|os.O_SYNC, 0666)
 	if err != nil {
-		return
+		return fmt.Errorf("failed opening %s: %w", fs.path, err)
 	}
 	// Ensure temp file is cleaned up
 	defer func() {
-		if e := os.RemoveAll(path); err == nil {
-			err = e
+		if e := os.RemoveAll(path); err == nil && e != nil {
+			err = fmt.Errorf("failed removing temporary file %s: %w", path, e)
+		}
+		if e := os.RemoveAll(fs.changeMgr.Path); err == nil && e != nil {
+			err = fmt.Errorf("failed removing saved field changes - %s: %w", fs.changeMgr.Path, e)
 		}
 	}()
+	fs.mu.RLock()
```

Comment:
This lets look-ups continue in the MeasurementFileSet.  Not sure (per earlier conversation) if there are any, but we are not changing the in-memory copy at all, so I see it as a Read operation.

Reply:
Ok - like I said I think this is fine (I can't think of any way it breaks at least), but I see the rename of the tmp file (which this lock is also protecting) as a write. But because of the read lock, any two threads that are concurrent here must both hold the read lock and so must be trying to write the same version of the MeasurementFieldSet.